src/map.rs

```diff
@@ -97,602 +101,755 @@ pub enum MapEvent<T: Tile, C: Chunk<T>> {
-    Some((sprite_rect.min.x() as usize, sprite_rect.min.y() as usize))
+/// A component bundle for `TileMap` entities.
+#[derive(Bundle, Debug)]
+pub struct TileMapComponents {
+    /// A `TileMap` which maintains chunks and its tiles.
+    pub tile_map: TileMap,
+    /// The transform location in a space for a component.
+    pub transform: Transform,
+    /// The global transform location in a space for a component.
+    pub global_transform: GlobalTransform,
 }
 
-fn set_tiles<T>(
-    tiles: &[T],
-    chunk_texture: &mut Texture,
-    sprite_sheet_texture: &Texture,
-    sprite_sheet_atlas: &TextureAtlas,
-    chunk_rect: Rect,
-    chunk_coord: Vec2,
-) where
-    T: Tile,
-{
-    let map_texture_size = chunk_texture.size.x() as usize;
-    let chunk_format_size = chunk_texture.format.pixel_size();
-    let format_size = chunk_texture.format.pixel_size();
-
-    let mut positions = Vec::with_capacity(tiles.len());
-    for tile in tiles.iter() {
-        if let Some(coord) = sprite_coord(tile, sprite_sheet_atlas) {
-            positions.push(coord);
-        }
-    }
-    positions.shrink_to_fit();
-    // We need this bool here to prevent borrow issues in the loop.
-    let is_single = positions.len() == 1;
-
-    let width = sprite_sheet_texture.size.x() as usize;
-    let rect_width = chunk_rect.width() as usize;
-    let rect_height = chunk_rect.height() as usize;
-    let rect_y = chunk_coord.y() as usize;
-    let rect_x = chunk_coord.x() as usize;
-    for (sprite_x, mut sprite_y) in positions.iter_mut() {
-        for bound_y in rect_y..rect_y + rect_height {
-            let begin = (bound_y * map_texture_size + rect_x) * chunk_format_size;
-            let end = begin + rect_width * chunk_format_size;
-            let sprite_begin = (sprite_y * width + *sprite_x) * format_size;
-            let sprite_end = sprite_begin + rect_width * format_size;
-            if is_single {
-                chunk_texture.data[begin..end]
-                    .copy_from_slice(&sprite_sheet_texture.data[sprite_begin..sprite_end]);
-            } else {
-                let data = &mut chunk_texture.data[begin..end];
-                let sprite_data = &sprite_sheet_texture.data[sprite_begin..sprite_end];
-                for x in 0..((sprite_end - sprite_begin) / format_size) {
-                    let inner_begin = x * format_size;
-                    let inner_end = inner_begin + format_size;
-                    let inner_data = &sprite_data[inner_begin..inner_end];
-                    if inner_data[3] != 0 {
-                        data[inner_begin..inner_end]
-                            .copy_from_slice(&sprite_data[inner_begin..inner_end])
-                    }
+/// The event handling system for the `TileMap` which takes the types `Tile`, `Chunk`, and `TileMap`.
+pub fn map_system(
+    mut commands: Commands,
+    mut chunks: ResMut<Assets<Chunk>>,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut query: Query<(&Entity, &mut TileMap)>,
```

Comment:
You're querying for `&Entity` rather than `Entity` which means this query never produces any results. Once you fix that things in the renderer start breaking because it actually has Chunks to try to render.

Reply:
Ah, just assumed it was &Entinty as others need a reference of some sort.